sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+
+Another mechanism could be that the compiler provide attributes which can be applied to the pointer declaration to convey some compile-time constant information. In this case users would change their code to:
+```cpp
+{
+  [[special_annotations::use_case_x_attribute(I)]] SomeType* a; // 'I' is the compiler time constant information being annotated on 'a'
+  ...
+  *a = some_val;
+  ...
+  read_val = *a
+}
+```
+However, this mechanism fails to meet requirements (2) and (3) listed above since:
+
+  * Adding new attributes for can add non-trivial attribute creation overheads for compiler developers. With growing set of attributes, there is an associated burden of maintaining them and of preparing diagnostic checks.
+  * Compilers are not required to preserve information propagated as attributes. Hence this method is only suitable for hints, and not functional directives. Further, the compiler may not be able to observe a particular pointer load/store call and successfully trace back to the declaration to deduce which annotation was applied to it.
+
+The `annotated_ptr` class described in this document is a templated type that encapsulates a pointer. The template accepts a list of compile-time constant properties. The implementation of the class defined here should preserve the information provided as compile-time constant properties through all uses of the pointer unless noted otherwise.
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification, Section 6.3.3 "Feature test macros". Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_ANNOTATED_PTR` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features
+that the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version
+|===
+
+=== Representation of `annotated_ptr`
+
+`annotated_ptr` is a class template, parameterized by the type of the underlying allocation `T`, and a list of properties `PropertyListT`.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = property_list<>>
+class annotated_ptr {
+  ...
+----
+
+`PropertyListT` enables properties to be associated with an `annotated_ptr`.
+Properties may be specified for an `annotated_ptr` to provide semantic
+modification or optimization hint information.
+
+Example uses of a property are:
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi;
+{
+  sycl::queue q;
+  annotated_ptr<int, property_list_t<align<4>>> kernel_arg; // alignment of the pointer in bytes specified using the property 'align'
```

Comment:
This example has UB because it dereferences an uninitialized pointer in the device code.  Could you change it as follows to make it possible defined:

```suggestion
  annotated_ptr<int, property_list_t<align<4>>> kernel_arg = ... // alignment of the pointer in bytes specified using the property 'align'
```